### ISSUE TITLE
Rewrite parsers in std to use `haxe.io.Input` instead of strings as data sources

### ISSUE THREAD

**Issue by RealyUniqueName** [HaxeFoundation/haxe#8085]
Since Haxe now supports multibyte strings by default, string-based parsers became slower on some targets (at least on php: #8083)
`haxe.xml.Parser`
`haxe.format.JsonParser`
`haxe.Template`
Anything else?

**Comment by ncannasse** [comment]
haxe.io.Input is maybe a bit too highlevel, calling readByte is a bit expensive and requires a dynamic dispatch. Maybe we should have something more like a StringReader that allows the operations we need in a parser in an optimized way, something that would replace StringTools.fastCodeAt/StringTools.isEOF but with a more high level API:

```
class StringReader {
public inline function new( s : String );
inline function nextChar() : Int; 
inline function isEof( c : Int ) : Bool;
inline function getCurrentPos() : Int;
inline function substr( startPos : Int, endPos : Int ) : String; 
}
```

Regarding parsers, you can search std library for `fastCodeAt`, you will find some occurrences in `haxe.io.Path` and also `haxe.Unserializer`

**Comment by Simn** [comment]
What would also help is having an `actuallyFastCodeAt` which doesn't come with the weird Eof specification. You have the string length, there's really no excuse for reading out of bounds...

**Comment by ncannasse** [comment]
Right, that could be replacing isEof(c:Int) by isEnd() which means we have reached the end of the string.

**Comment by Simn** [comment]
This was closed in #9467.
